File: dsa2/tp/tp01/Is/Main.c
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>

#define BUFFER_SIZE 512

char* toLowercase(char entry[]){ // FREE!!! FREE!!! // FREE!! // FREE!! FREE!!! FREE!!!!!! FREE!!!!!1!
  char* resultado = (char*)malloc(BUFFER_SIZE);
  int i;
  for(i=0; entry[i] != '\0'; i++){
      if(entry[i] >= 'A' && entry[i] <= 'Z'){
        resultado[i] = (char)(entry[i] + 32);
      }else if(entry[i] == '\n'){
        break;
      }else{
        resultado[i] = entry[i];
      }
  }
  resultado[i] = '\0';
  return resultado;
}

bool isVowel(char entry[], int i){
  if(entry[i] == '\0'){
    return true;
  }else if(entry[i] != 'a' && entry[i] != 'e' && entry[i] != 'i'  && entry[i] != 'o' && entry[i] != 'u'){
    return false;
  }else{
    return isVowel(entry, i+1);
  }
}


bool isConsonant(char entry[], int i){
  if(entry[i] == '\0'){
    return true;
  }else if(!((entry[i] >= 'a' && entry[i] <= 'z') && (entry[i] != 'a' && entry[i] != 'e' && entry[i] != 'i' && entry[i] != 'o'  && entry[i] != 'u'))){
    return false;
  }else{
    return isConsonant(entry, i+1);
  }
}

bool isInt(char entry[], int i){
  if(entry[i] == '\0'){
    return true;
  }else if(!((entry[i] >= '0' && entry[i] <= '9') ||  (entry[i] == '-' && i == 0))){
    return false;
  }else{
    return isInt(entry, i+1);
  }
}

bool isReal(char entry[], int i, int dotCount){
  if(entry[i] == '.' || entry[i] == ',') dotCount++;

  if(entry[i] == '\0'){
    if(dotCount == 1 || dotCount == 0){
      return true;
    }else{
      return false;
    }
  }else if(!((entry[i] >= '0' && entry[i] <= '9') || (entry[i] == '-' && i == 0) || (entry[i] == ',' && dotCount == 1) || (entry[i] == '.' && dotCount == 1))){
  return false;
  }else{
    return isReal(entry, i+1, dotCount);
  }
}

bool Vowel(char entry[]){
  return isVowel(entry, 0);
}

bool Consonant(char entry[]){
  return isConsonant(entry, 0);
}

bool Int(char entry[]){
  return isInt(entry, 0);
}

bool Real(char entry[]){
  return isReal(entry, 0, 0);
}
  

int main(){
  char buffer[BUFFER_SIZE];

  while(fgets(buffer, BUFFER_SIZE, stdin) != NULL){
    if(buffer[0] == 'F' && buffer[1] == 'I' && buffer[2] == 'M') break;
    char* entry = toLowercase(buffer);

    if(Vowel(entry)){
      printf("SIM ");
    }else{
      printf("NAO ");
    }

    if(Consonant(entry)){
      printf("SIM ");
    }else{
      printf("NAO ");
    }

    if(Int(entry)){
      printf("SIM ");
    }else{
      printf("NAO ");
    }

    if(Real(entry)){
      printf("SIM\n");
    }else{
      printf("NAO\n");
    }

    free(entry);
  }
  return 0;
}

File: dsa2/tp/tp01/Is/Main.java
import java.util.Scanner;

class Main{
  public static String toLowercase(String entry){
    String resultado = "";
    for(int i=0; i<entry.length(); i++){
      if(entry.charAt(i) >= 'A' && entry.charAt(i) <= 'Z'){
        resultado+= (char)(entry.charAt(i) + 32);
      }else{
        resultado+= entry.charAt(i);
      }
    }
    return resultado;
  }

  public static String isVowel(String entry){
    for(int i=0; i<entry.length(); i++){
        if(entry.charAt(i) != 'a' && entry.charAt(i) != 'e' && entry.charAt(i) != 'i' && entry.charAt(i) != 'o' && entry.charAt(i) != 'u') return "NAO";
    }
    return "SIM";
  }

  public static String isConsonant(String entry){
    for(int i=0; i<entry.length(); i++){
      if(!((entry.charAt(i) >= 'a' && entry.charAt(i) <= 'z') && (entry.charAt(i) != 'a' && entry.charAt(i) != 'e' && entry.charAt(i) != 'i' && entry.charAt(i) != 'o' && entry.charAt(i) != 'u'))){
        return "NAO";
      }
    }
    return "SIM";
  }

  public static String isInt(String entry){
    for(int i=0; i<entry.length(); i++){
      if(!((entry.charAt(i) >= '0' && entry.charAt(i) <= '9') || (entry.charAt(i) == '-' && i == 0))) return "NAO";
    }
    return "SIM";
  }

  public static String isReal(String entry){
    int dotCount = 0;
    for(int i=0; i<entry.length(); i++){
      if(entry.charAt(i) == '.' || entry.charAt(i) == ',') dotCount++;
      if(!((entry.charAt(i) >= '0' && entry.charAt(i) <= '9') || (entry.charAt(i) == '-' && i == 0) || (entry.charAt(i) == ',' && dotCount == 1) || (entry.charAt(i) == '.' && dotCount == 1))) return "NAO";
    }
    if(dotCount == 1 || dotCount == 0){
      return "SIM";
    }else{
      return "NAO";
    }
  } 

  public static void main(String[] args){
      Scanner sc = new Scanner(System.in);

      while(sc.hasNextLine()){
        String entry = sc.nextLine();
        if(entry.length() == 3 && entry.charAt(0) == 'F' && entry.charAt(1) == 'I' && entry.charAt(2) == 'M') break;
        entry = toLowercase(entry);
        System.out.println(isVowel(entry) + " " + isConsonant(entry) + " " + isInt(entry) + " " + isReal(entry));
      }
      
      sc.close();
  } 
}
